src/HomologicalAlgebra/BoundedCocomplexes.jl

```diff
@@ -0,0 +1,397 @@
+# zeroes. Example:
+#
+# math. index:   -5   -4   -3   -2   -1    0    1    2    3
+#
+# groups:         0    X⁻⁴→ X⁻³→ X⁻²→ X⁻¹→ X⁰→  X¹   0    0
+#
+# intern. index:  -    1    2    3    4    5    6    -    - 
+# 
+# Internally, the groups and morphisms are stored in vectors 
+# starting with the `internal index` 1. For the translation from 
+# the mathematical to the internal index, we also store a shift, 
+# which is 5 in the above case.
+#
+########################################################################
+
+mutable struct BoundedCocomplex{GroupType, GroupHomType} <: ChainCocomplex{GroupType, GroupHomType}
+  shift::Int
+  C::Vector{GroupType}
+  d::Vector{GroupHomType}
+
+  # fiends for caching
+  Z::Vector
+  B::Vector
+  H::Vector
+
+  function BoundedCocomplex(C::Vector{GroupType}, d::Vector{GroupHomType}; shift::Int=1, check::Bool=false) where {GroupType, GroupHomType}
+    n = length(C)
+    n > 0 || error("need at least one instance of an object")
+    length(d) == n-1 || error("number of maps does not coincide with the number of groups")
+    for i in 1:n-1
+      domain(d[i]) == C[i] || error("domain of definition does not coincide with the given element")
+      codomain(d[i]) == C[i+1] || error("codomain of definition does not coincide with the given element")
+    end
+    if check
+      for i in 1:n-2
+	iszero(compose(d[i], d[i+1])) || error("composition of the differentials does not equate to zero")
+      end
+    end
+
+    return new{GroupType, GroupHomType}(shift, C, d)
+  end
+end
+
+function getindex(C::BoundedCocomplex, i::Int)
+  if i+shift(C) < 1 
+    return zero_object(first(C))
+  end
+  if i+shift(C)> length(cochains(C))
+    return zero_object(last(C))
+  end
+  return C.C[i+C.shift]
+end
+
+differentials(C::BoundedCocomplex) = C.d
+
+function differential(C::BoundedCocomplex, i::Int) 
+  if i+shift(C) < 1 || i+shift(C) >= length(cochains(C))
+    return zero_morphism(C[i], C[i+1])
+  end
+  return C.d[i+C.shift]
+end
+
+length(C::BoundedCocomplex) = length(C.C)
+last(C::BoundedCocomplex) = last(C.C)
+first(C::BoundedCocomplex) = C.C[1]
+shift(C::BoundedCocomplex) = C.shift
+shift(C::BoundedCocomplex, i::Int) = BoundedCocomplex(cochains(C), differentials(C), shift=shift(C)+i, check=false)
+start_index(C::BoundedCocomplex) = -shift(C)+1
+end_index(C::BoundedCocomplex) = start_index(C) + length(C) - 1
+
+cochains(C::BoundedCocomplex) = C.C
+
+function kernels(C::BoundedCocomplex)
+  if !isdefined(C, :Z)
+    C.Z = [kernel(differential(C,i)) for i in start_index(C):end_index(C)]
+  end
+  return C.Z
+end
+
+function boundaries(C::BoundedCocomplex)
+  if !isdefined(C, :B)
+    C.B = [image(differential(C,i)) for i in start_index(C)-1:end_index(C)-1]
+  end
+  return C.B
+end
+
+function cohomologies(C::BoundedCocomplex)
+  if !isdefined(C, :H)
+    C.H = Vector()
+    Z = kernels(C)
+    B = boundaries(C)
+    for i in 1:length(C)
+      push!(C.H, quo(Z[i], B[i]))
+    end
```

Comment:
This part of the code is not yet functional. I think we will have to agree on a general set of (internal) functions to compute cohomology throughout Oscar. Otherwise, one would have to write different code here depending on whether one has Abelian groups from Hecke, Vector spaces, Modules, or whatever.